docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
-existing `_ss_environment.php` file.
+Take for example the following SilverStripe 3 `composer.json` file.
+```json
+{
+    "name": "acme/cms-website",
+    "description": "The ACME website project.",
+    "license": "BSD-3",
+    "require": {
+        "php": ">=5.3.3",
+        "silverstripe/cms": "3.6.5@stable",
+        "silverstripe/framework": "3.6.5@stable",
+        "silverstripe/reports": "3.6.5@stable",
+        "silverstripe/siteconfig": "3.6.5@stable",
+        "dnadesign/silverstripe-elemental": "^1.8.0"
+    }
+}
+```
 
+You can upgrade the `composer.json` file with this command:
+```bash
+/var/www/SS_acme > upgrade-code recompose --recipe-core-constraint=1.1
 ```
-cd ~/my-project-root
-upgrade-code environment --write
+
+The `--recipe-core-constraint` flag can be use to target a specific version of `silverstripe/recipe-core`. If this flag is omitted, the project will be upgraded to the latest stable version. You can use the `--strict` option if you want to use more conservative version constraints. If you want to save your changes, use the `--write` flag,
+
+Your upgraded `composer.json` file will look like this.
+```json
+{
+    "name": "acme/cms-website",
+    "description": "The ACME website project.",
+    "license": "BSD-3",
+    "require": {
+        "dnadesign/silverstripe-elemental": "^2.1",
+        "php": ">=5.6",
+        "silverstripe/recipe-cms": "^1.1"
+    }
+}
 ```
 
-Read the [upgrader `environment` command documentation](https://github.com/silverstripe/silverstripe-upgrader/blob/master/docs/en/environment.md)
-for more details.
+If the `recompose` command can't find a SilverStripe 4 compatible version for one of your package, it will keep this dependency in your `composer.json` file with its existing constraint.
+
+### Manually upgrading your dependencies
+The instruction in this section assumed you'll be editing your `composer.json` file in a text editor.
+
+#### Switching to _recipes_
+Where possible, we recommend you use recipes.
+
+If your SilverStripe 3 project requires the `silverstripe/cms` package, replace that dependency with `silverstripe/recipe-cms`. Set the version constraint for `silverstripe/recipe-cms` to:
+* `~1.0.0` to upgrade to SilverStripe 4.0
+* `~1.1.0` to upgrade to SilverStripe 4.1
+* `~1.2.0` to upgrade to SilverStripe 4.2
+* and so on.
+
+If your SilverStripe 3 project requires the `silverstripe/framework` package without `silverstripe/cms`, replace `silverstripe/framework` with `silverstripe/recipe-core`. Set the version constraint for `silverstripe/recipe-core` to:
+* `~1.0.0` to upgrade to SilverStripe 4.0
+* `~1.1.0` to upgrade to SilverStripe 4.1
+* `~1.2.0` to upgrade to SilverStripe 4.2
+* and so on.
+
+The following packages are implicitly required by `silverstripe/recipe-core`. They can be removed from your `composer.json` dependencies if you are using `silverstripe/recipe-core` or `silverstripe/recipe-cms`.
+* `sivlerstripe/framework`
+* `silverstripe/config`
```

Comment:
It's a lie that tells the truth. Basically I want to communicate that if you use the recipe, you don't need to explicitly install these packages. 

I don't really want to start explaining to people, "this package and this package didn't exist in SS3 because they were part of this other package that got split up, so you won't have them in you SS3 composer file, however if you want to manually define everything you need to include them".

Also, it's not strictly speaking a SS3 to SS4 upgrade guide. In theory you could follow this guide to upgrade from SS4.0 to SS4.2.